modules: update yaksa

raffenet opened the pull request:
## Pull Request Description

pmodels/yaksa#213: backend/ze: fix warnings about unused labels
pmodels/yaksa#214: cuda: Delay stream creation
pmodels/yaksa#215: cuda: Allow configure to continue after nvcc check
pmodels/yaksa#216: cuda: delay query p2p accessibility

Fixes pmodels/mpich#5810

## Author Checklist
* [ ] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [ ] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [ ] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [ ] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

raffenet commented:
test:mpich/ch4/gpu

raffenet commented:
Failures in the ch4:ofi configuration
```
not ok 1949 - ./rma/accfence1 4
  ---
  Directory: ./rma
  File: accfence1
  Num-procs: 4
  Timeout: 600
  Date: "Fri Mar 18 12:13:35 2022"
  ...
## Test output (expected 'No Errors'):
## Abort(605728271) on node 0 (rank 0 in comm 0): Fatal error in internal_Win_create: Other MPI error, error stack:
## internal_Win_create(106)..........: MPI_Win_create(base=0x7fb434000000, size=268435456, disp_unit=4, MPI_INFO_NULL, comm=0x84000002, win=0x7ffd9ab7dbdc) failed
## MPID_Win_create(79)...............: 
## MPIDIG_mpi_win_create(889)........: 
## MPIDI_SHM_mpi_win_create_hook(114): 
## MPIDI_IPC_mpi_win_create_hook(85).: 
## MPIDI_GPU_get_ipc_attr(147).......:  gpu_get_buffer_info failed
```
The error is coming from `cuMemGetAddressRange` and is value 500.
> CUDA_ERROR_NOT_FOUND = 500
> This indicates that a named symbol was not found. Examples of symbols are global/constant variable names, driver function names, texture names, and surface names.
This happening during window creation, which is done using a buffer on the second GPU on the node. Unclear why it only happens in this test and not the other window tests. Maybe because memtypes are random?

raffenet commented:
@jdinan @jirikraus any clue why `cuMemGetAddressRange` might fail with the change to lazily allocate CUDA resources during `MPI_Init`?

raffenet commented:
FWIW, the failures in the ch4:ucx configuration are consistent with our nightly tests, so can be ignored for this PR.

raffenet commented:
> This is happening during window creation, which is done using a buffer on the second GPU on the node. Unclear why it only happens in this test and not the other window tests. Maybe because memtypes are random?

Presumably the CUDA runtime is not initialized which is causing the failure with the driver API call. The runtime initialization is implicit, so we need to find a way to ensure it happens before this stage. Though by the time of the failure the test has allocated device memory and MPICH has queried the cuda runtime to detect whether the pointer lives on the device.

The docs speak specifically to mixing runtime and driver APIs like we are doing https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#interoperability-between-runtime-and-driver-apis.

jirikraus commented:
> @jdinan @jirikraus any clue why `cuMemGetAddressRange` might fail with the change to lazily allocate CUDA resources during `MPI_Init`?

To clarify what is happening. You are calling `cuMemGetAddressRange` on a pointer referencing GPU memory on the GPU controlled by another MPI rank on the same node and this pointer was imported into the current MPI process with CUDA IPC? 

cc @Akshay-Venkatesh for visibility and in case he has immediate input on this.

raffenet commented:
> > > > @jdinan @jirikraus any clue why `cuMemGetAddressRange` might fail with the change to lazily allocate CUDA resources during `MPI_Init`?
> > > 
> > > 
> > > To clarify what is happening. You are calling `cuMemGetAddressRange` on a pointer referencing GPU memory on the GPU controlled by another MPI rank on the same node and this pointer was imported into the current MPI process with CUDA IPC?
> > > cc @Akshay-Venkatesh for visibility and in case he has immediate input on this.
> > 
> > 
> > The error occurs when calling `cuMemGetAddressRange` on a GPU memory pointer allocated by the same process. Both GPUs in the system are shared by the 2 MPI ranks in the example. No IPC handles have been opened/exchanged at the time of the error.
> 
> Thanks for clarifying. So each process controls two GPUs? Why is this the case? Is this something you intent to support? So far my understanding is that each MPI rank is bound to a single GPU and that affinity is immutable for the lifetime of the application.

In this instance our test case is allocating memory on each device and verifying that it is accessible from other processes. It is not a configuration we would recommend, but I think its one we do support. In other words, we won't stop users from doing such things 😅 .

raffenet commented:
> I just did a quick standalone test and `cuMemGetAddressRange` fails with error `500` for me if the passed in `CUdeviceptr` is not on the current device. E.g. if I allocated on GPU 0 then call `cudaSetDevice(1)` `cuMemGetAddressRange` fails with error `500` for me on the allocation. Maybe lazy initializing and switching between devices caused a resources to end up at a different GPU compared to initializing everything during `MPI_Init`. @raffenet Can you confirm somehow that this is the root cause of the error you are seeing?

Thanks for the tip! I can confirm that when current device is set to 0 and we attempt a pointer query on a device 1 allocation, error 500 occurs. At that point in the program, the only thing we have done with device 1 was allocate a buffer. If I add a query function right after allocating the buffer in each process, then the 500 error goes away.

To summarize, the failure occurs when:
```
cudaSetDevice(1);
cudaMalloc(&buffer, size);
cudaSetDevice(0);
cudaPointerGetAttributes(&attr, buffer);
```
It goes away if we do:
```
cudaSetDevice(1);
cudaMalloc(&buffer, size);
cudaPointerGetAttributes(&attr, buffer);
cudaSetDevice(0);
cudaPointerGetAttributes(&attr, buffer);
```

I think inserting a dummy pointer query after allocation in the test harness might be a reasonable fix. @hzhou what do you think?

hzhou commented:
> To summarize, the failure occurs when:
> 
> ```
> cudaSetDevice(1);
> cudaMalloc(&buffer, size);
> cudaSetDevice(0);
> cudaPointerGetAttributes(&attr, buffer);
> ```
> 
> It goes away if we do:
> 
> ```
> cudaSetDevice(1);
> cudaMalloc(&buffer, size);
> cudaPointerGetAttributes(&attr, buffer);
> cudaSetDevice(0);
> cudaPointerGetAttributes(&attr, buffer);
> ```
> 
> I think inserting a dummy pointer query after allocation in the test harness might be a reasonable fix. @hzhou what do you think?

User allocates the memory, right? We can't always ensure to query on the same device. Afterall, we need `cudaPointerGetAttributes` to get the device. Am I missing something?

raffenet commented:
I edited my example above because I meant to use `cuMemGetAddressRange` like we do in the MPI library.

hzhou commented:
> I edited my example above because I meant to use `cuMemGetAddressRange` like we do in the MPI library.

I see. So maybe we can run a `cudaPointerGetAttributes`, potentially set the right device, before we call `cuMemGetAddressRange`?

raffenet commented:
> > I edited my example above because I meant to use `cuMemGetAddressRange` like we do in the MPI library.
> 
> I see. So maybe we can run a `cudaPointerGetAttributes`, potentially set the right device, before we call `cuMemGetAddressRange`?

Yes, that might work. Let me give it a try.

raffenet commented:
test:mpich/ch4/gpu
testlist: rma

raffenet commented:
test:mpich/ch4/gpu
testlist: rma

Akshay-Venkatesh commented:
@raffenet I don't understand why 
```
cudaSetDevice(1);
cudaMalloc(&buffer, size);
cudaPointerGetAttributes(&attr, buffer);
cudaSetDevice(0);
cuMemGetAddressRange(..., buffer);
```
should work. Are there any places where p2p is explicitly enabled/disabled (say using [cudaDeviceDisablePeerAccess](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__PEER.html#group__CUDART__PEER_1g9663734ad02653207ad6836053bf572e)). Along the lines of what @jirikraus mentioned, if thread is bound to a specific device context (this happens implicitly when calling cudaSetDevice), then in general it is not permitted to access buffers allocated on different device (unless the two devices are peer-accessible). I don't see how calling pointer query API solves the problem. Even if there is side-effect of the error going away, it's probably better to set the right context on the calling thread (either by cudaSetDevice or using drive API to set current context) and then access the buffer in question. Lmk if I've missed anything.

raffenet commented:
test:mpich/ch4/gpu
testlist: rma

raffenet commented:
test:mpich/ch4/gpu

raffenet commented:
test:mpich/ch4/gpu

hzhou commented on src/mpid/ch4/shm/ipc/gpu/gpu_init.c:
```diff
@@ -43,10 +43,10 @@ static void ipc_handle_free_hook(void *dptr)
 
     if (MPIR_CVAR_CH4_IPC_GPU_HANDLE_CACHE) {
-        mpl_err = MPL_gpu_get_buffer_bounds(dptr, &pbase, &len);
+        MPIR_GPU_query_pointer_attr(dptr, &gpu_attr);
```
I would do the pointer query inside `MPL_gpu_get_buffer_bounds`. It might be cuda version specific, so it makes sense to do it inside MPL wrappers.

raffenet replied:
If we do the query in `MPL_gpu_get_buffer_bounds`, the caller might end up doing a redundant query. That is unless we modify the API to also return the device handle or pointer attributes. But if we return it through the API and the backend didn't require setting the device, we could end up doing the query for no reason.

Right now the 2 call sites in MPICH are already doing a pointer attribute query, so whichever design we choose is kind of moot.

raffenet commented:
test:mpich/ch4/gpu

raffenet commented:
Removed testing hack patch. Dropped MPL API modification and added pointer attribute query within `MPL_gpu_get_buffer_bounds` as suggested.
test:mpich/ch4/gpu
